### ISSUE TITLE
Add S3-backed filesystem tools (ApplyDiff/ListFiles/ReadFile/DeleteFile) via AgentCore Gateway

### ISSUE THREAD

**Issue by waltmayf** [waltmayf/agents4energy#240]
## Goal

Give the chat harness agent (`MyHarness`) an ergonomic, S3-backed **filesystem** it can read and write, exposed as four tools through the **AgentCore Gateway** (as a Lambda-backed gateway target):

- **`ApplyDiff`** — create or modify a file
- **`ListFiles`** — list files/prefixes ("directory listing")
- **`ReadFile`** — read a file's contents
- **`DeleteFile`** — delete a file

This lets an agent read shared domain documentation (e.g. `/docs/production/gas_lift.md`) and write business outputs (reports, analytic outputs) into a shared file space, treating S3 like a local filesystem.

This is the first Lambda-backed gateway target in the repo, and the first use of Amplify Storage.

---

## Filesystem model (S3-as-filesystem)

One bucket (Amplify Storage), **one root prefix `files/`** that every path branches from. This keeps path handling uniform — there is no per-prefix special-casing.

- **Absolute paths** (leading `/`) resolve from the filesystem root. `/docs/production/gas_lift.md` → S3 key `files/docs/production/gas_lift.md`.
- **Relative paths** (no leading `/`) resolve from the same root. `reports/q3.md` → S3 key `files/reports/q3.md`. (There is no per-session working directory — see note below.)
- Normalize `.`/`..` safely and **reject any path that escapes the `files/` root** (path-traversal guard). This is the only guard.

**Conventions (not enforced):**
- Seeded/shared domain documentation lives under `files/docs/...` by convention. It is **not** a separate enforced prefix and is **not** read-only — the tools can read and write anywhere under `files/`. (Seeded docs are re-seeded on deploy, so an accidental overwrite is recoverable.)
- Business outputs, reports, and scratch files can live anywhere under `files/`, e.g. `files/reports/...`.

**No per-session isolation (deliberately dropped).** All authenticated users / all chat sessions share the same `files/` space. Per-session working directories were considered and dropped: the AgentCore Gateway invokes a Lambda target with **only** `event` (the tool's argument values) and a fixed `context` (gateway/target/tool ids + per-request ids) — it does **not** forward the harness `runtimeSessionId`, the caller's Cognito identity, or inbound request headers to a Lambda target (confirmed against the AWS "Lambda function targets" contract). Threading a session id would require either making session id an explicit tool argument or building a REQUEST interceptor + HTTP MCP-server target — both out of scope here. If per-session sandboxing is wanted later, open a follow-up issue.

---

## Tool specifications

All tools take paths per the resolution rules above. Return structured, model-friendly results (and actionable errors).

### `ApplyDiff` — create or modify a file

Accept the **SEARCH/REPLACE block format** (the Aider / Cline / Roo-Code `apply_diff` style), applied **server-side with fuzzy matching**. This format is chosen deliberately: it is the most prevalent file-edit format in open-source-agent training data (so both `openai.gpt-oss-120b` and Claude produce it reliably), it keys off **content not line numbers** (models are unreliable with line numbers), and it is simple to apply server-side. **Do NOT use unified diff (`diff -u`/`@@ -n,m` hunks)** — models get the line-number math wrong and applies fail.

**Args:**
- `path` (string, required) — target file, per resolution rules.
- `diff` (string, required) — one or more SEARCH/REPLACE blocks.

**Block grammar:**
```
<<<<<<< SEARCH
:start_line:<N>          (optional hint; the applier does NOT trust it)
-------
<exact existing content, including whitespace/indentation>
=======
<replacement content>
>>>>>>> REPLACE
```

**Application rules to implement:**
- Fixed markers: `<<<<<<< SEARCH`, optional `:start_line:<N>` followed by `-------` header, `=======` divider, `>>>>>>> REPLACE`. Tolerate a stray trailing `>` after `SEARCH`.
- Multiple blocks per call, applied top-to-bottom against the evolving buffer.
- Matching: try exact match first; if none, fuzzy match via normalized Levenshtein similarity (default threshold exact `1.0`, allow lowering to ~`0.9`). Search near `:start_line:` first, then middle-out. Normalize smart quotes and line endings before comparing; preserve the file's original whitespace on replace.
- **File creation:** an **empty SEARCH block** targeting a non-existent `path` creates the file with the REPLACE body as its full content. (Also acceptable: a full-content write when the file doesn't exist.)
- Errors must be actionable and let the model self-correct: `"no match found for SEARCH block N"` or `"SEARCH block N matched M locations — add more surrounding context"` (mirror Anthropic `str_replace` uniqueness guidance).
- Literal marker lines inside content are backslash-escaped.

Reference sources for the format/fuzzy-apply: Aider edit-formats docs (`aider.chat/docs/more/edit-formats.html`), Aider unified-diff analysis (`aider.chat/docs/unified-diffs.html`), Roo-Code `src/core/diff/strategies/multi-search-replace.ts`.

### `ListFiles`
- Args: `path` (string, optional — a prefix/"directory"; defaults to the `files/` root), `recursive` (boolean, optional, default false).
- Returns a listing distinguishing files from sub-"directories" (common prefixes), with sizes. Uses `ListObjectsV2` with `Delimiter: '/'` for non-recursive.

### `ReadFile`
- Args: `path` (string, required).
- Returns file contents as text. Guard size (e.g. cap at a sane max, return a clear error if exceeded). Handle not-found with a clear error.

### `DeleteFile`
- Args: `path` (string, required).
- Deletes the object. Return success/clear error (including a clear error if the file does not exist).

Keep tool names and `inputSchema` shapes exactly as the `ToolDefinition[]` inline `toolSchema` on the gateway target (see wiring below), so the harness sees `ApplyDiff`, `ListFiles`, `ReadFile`, `DeleteFile`.

---

## Implementation wiring (end-to-end — all of this is in scope)

Ground truth for how the gateway/targets/harness fit together is in `docs/agentic-architecture.md` and `web/amplify/backend.ts`. Follow existing patterns.

1. **Amplify Storage bucket.** Add `web/amplify/storage/resource.ts` with `defineStorage` (name e.g. `agentFiles`), and register it in `defineBackend({...})` in `web/amplify/backend.ts`. The repo has no storage today — this is new. All tool I/O is confined to the `files/` prefix within this bucket.

2. **S3-tools Lambda.** New function `web/amplify/functions/s3-tools/` (`resource.ts` + `handler.ts`), modeled on `web/amplify/functions/register-mcp-target/`. The handler is the MCP tool implementation for all four tools; **dispatch on the tool name from the gateway context** — `context.bedrockAgentCoreToolName` has the form `${target_name}__${tool_name}`, and the tool arguments arrive as the `event` object (a flat map of the tool's `inputSchema` args to values). Env: bucket name, region. Grant it `s3:GetObject`, `s3:PutObject`, `s3:DeleteObject`, `s3:ListBucket` scoped to the bucket ARN + `/files/*` (via `backend.<fn>.resources.lambda as LambdaFunction` + `.addToRolePolicy(new PolicyStatement(...))`, the pattern already used at `backend.ts` ~486-499). Give it the gateway invoke permission via `Function.addPermission('AllowGatewayInvoke', { principal: ServicePrincipal('bedrock-agentcore.amazonaws.com'), action: 'lambda:InvokeFunction', sourceArn: AGENTCORE_GATEWAY_ARN })` — mirror the existing `AllowGatewayInvoke` block at `backend.ts` ~816-822.

3. **Register the Lambda as a gateway target.** Use `CreateGatewayTargetCommand` with a **Lambda** target config (the installed `@aws-sdk/client-bedrock-agentcore-control` ^3.1071 supports it):
   ```ts
   targetConfiguration: {
     mcp: {
       lambda: {
         lambdaArn: S3_TOOLS_LAMBDA_ARN,
         toolSchema: { inlinePayload: [ /* ApplyDiff, ListFiles, ReadFile, DeleteFile ToolDefinitions */ ] },
       },
     },
   }
   ```
   Each `ToolDefinition` = `{ name, description, inputSchema: SchemaDefinition }` (SchemaType enum: OBJECT/STRING/ARRAY/BOOLEAN/INTEGER/NUMBER). Registration must be **idempotent** (create-if-absent — list existing targets by name first, or catch already-exists). Prefer doing this in a deploy-time step consistent with how the repo registers other agentcore resources; if a custom resource is needed, model it on the existing `E2eTestUser` custom resource (`web/amplify/constructs/e2eTestUser/`). Do NOT hand-edit `agentcore.json` targets (it's `targets: []` by design — targets are created at runtime).

4. **Expose the gateway to an agent as an `McpServer` row.** The harness only calls tools that arrive via `buildTools()` from an `McpServer` DynamoDB record linked to the agent. Create (idempotently) an `McpServer` record with `serverType: 'agentcore'` and `url = <gateway endpoint>` (`https://<gatewayId>.gateway.bedrock-agentcore.us-east-1.amazonaws.com/mcp`, derived from SSM `/agentcore/<stackName>/gateway_endpoint` — do not hardcode), and link it to a test agent via an `AgentMcpServer` join row to prove the path end-to-end. (This "gateway-as-McpServer" wiring does not exist yet in the repo.)

---

## Docs & tests

- Update `docs/agentic-architecture.md` (the MCP Tools section) to describe the new Lambda gateway target and the S3 filesystem model.
- Add a short new doc under `docs/` describing the agent filesystem (the single `files/` root, path rules, the four tools, the ApplyDiff format) — useful for both devs and users.
- Add unit tests for the pure logic: path resolution (absolute and relative both resolve under `files/`; `.`/`..` normalization; traversal rejection for paths that escape `files/`) and the SEARCH/REPLACE parser + fuzzy applier (exact match, fuzzy match, no-match error, multi-match error, multi-block, create-via-empty-SEARCH). These don't need AWS.
- `npx tsc --noEmit` must pass before pushing.

---

## Out of scope (do NOT build here)

- Seeding the artificial-lift agent record and uploading the actual `files/docs/...` domain content — that's a **follow-up** issue. This issue delivers the *capability* (bucket + four tools + wiring) and can link the McpServer to an existing test agent to prove the path works end-to-end.
- Per-session and per-user isolation. All sessions/users share the single `files/` space; there is no working-directory concept. (Can be revisited in a follow-up.)

---

## Dispatch / delivery constraints

- Remote runs have a hard ~3h ceiling and **lose all unpushed work**. Push a **draft PR as soon as the code type-checks** (`npx tsc --noEmit`), then keep iterating on that branch. Do not wait until everything is done to push.
- Keep the PR focused on this issue.

The PR that resolves this issue must include `Closes #240` in its body.

**Comment by waltmayf** [comment]
@agentcore-claude Please implement this issue end-to-end following the spec above.

Key reminders:
- Push a **draft PR as soon as `npx tsc --noEmit` passes**, then keep iterating on that branch — runs have a hard ~3h ceiling and any unpushed work is lost.
- The PR body **must** include `Closes #240` on its own line so the issue auto-closes on merge.
- Verify the session-CWD approach empirically (header forwarding vs explicit `sessionId` arg) before committing to it, and document your choice in the PR.
- Run `npx tsc --noEmit` before every push; keep `docs/` updated.

**Comment by waltmayf** [comment]
@agentcore-claude **Spec simplified — please re-read the issue body before/while building.** Key changes from the original dispatch:

- **Single root prefix `files/`** (not two prefixes). Absolute `/docs/...` → `files/docs/...`; relative `reports/q3.md` → `files/reports/q3.md`. Only guard is path-traversal (cannot escape `files/`).
- **No `docs/` read-only enforcement** — the agent may read/write anywhere under `files/`. Seeded docs live at `files/docs/...` by convention only.
- **Per-session CWD is DROPPED entirely.** No `sessionId` threading, no header injection, no `workspace/id=<sessionId>/`. A Lambda gateway target receives only the tool args (`event`) + fixed gateway metadata (`context`) — no session id — so this was cut for simplicity. All sessions share one `files/` space.

The rest (Amplify Storage bucket, four tools, Lambda gateway target, McpServer wiring, ApplyDiff SEARCH/REPLACE format) is unchanged. If you already started on the old design, discard the prefix/CWD machinery.

**Comment by waltmayf** [comment]
@agentcore-claude PR #241 is close — the core logic (path resolver, diff applier, tests) is solid and the spec was followed correctly. But code review found **two deploy-blocking issues** (both invisible to CI, since CI does not deploy). Please fix them on the same branch (`feature/s3-filesystem-tools-240`) and push:

**B1 (blocking) — Gateway role missing `lambda:InvokeFunction` on the s3-tools Lambda.** `backend.ts` adds only the resource-based `AllowGatewayInvoke` policy. As your own PR description documents, the gateway's **execution role** also needs an identity-based `lambda:InvokeFunction` grant on the target Lambda ARN — the two IAM checks are separate. The vendored `@aws/agentcore-cdk` `Gateway` construct auto-adds this (`lambda:InvokeFunction` on `[lambdaArn, lambdaArn+":*"]`) only when it creates the Lambda target itself; because you register the target out-of-band via the custom resource, that auto-grant never runs, so `CreateGatewayTarget` will fail synchronous validation (400) and roll back. Fix: expose the gateway execution role from `AgentCoreApplication` (e.g. `addGatewayRolePolicy(name, stmt)` delegating to the `AgentCoreMcp` construct's `this.role`) and grant `lambda:InvokeFunction` on `[s3ToolsLambda.functionArn, s3ToolsLambda.functionArn + ":*"]`. Keep the resource-based policy too.

**B2 (blocking) — MCP-server seed will get `Unauthorized`.** The seed signs AppSync with the Lambda IAM role (SigV4), but `Agent`/`McpServer`/`AgentMcpServer` authorize `authenticated() + owner()` only. The `active-run.js` pattern you copied works only because `ChatSession`/`ActiveRun` also declare `allow.guest()`, which admits IAM-signed requests in identityPool mode. Add `allow.guest()` (or an explicit `allow.resource(<seed fn>)` grant) to those three models in `agentConfig.schema.ts`, and confirm an owner()-model create succeeds with no Cognito sub and the demo row is readable by end users.

**Should-fix (fold in):**
- S1: `s3ToolsMcpServerSeed/resource.ts` grants `appsync:GraphQL` on `apis/*` (all APIs) — scope to the concrete API ARN (`backend.data.resources.cfnResources.cfnGraphqlApi.attrApiId`).
- S2: `s3ToolsGatewayTarget/resource.ts` grants `bedrock-agentcore:*GatewayTarget` on `*` — scope to the gateway ARN where the action allows resource-level perms.
- S3: seed matches `McpServer` by `url == gatewayEndpoint`; matching by a stable name/slug avoids orphaning the old row if the gateway endpoint changes.

Please **re-verify B1 and B2 against the live deployed gateway** after fixing (deploy the sandbox and confirm `CreateGatewayTarget` succeeds and the seed row is created) — these are deploy-time failures CI does not exercise. Push to the existing PR branch; keep `Closes #240` in the PR body. Full detail is in the request-changes review on PR #241.
